index.go

```diff
@@ -0,0 +1,275 @@
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (i Index) GetAllUpdateStatus() (resp []Update, err error) {
+	resp = []Update{}
+	req := internalRequest{
+		endpoint:            "/indexes/" + i.UID + "/updates",
+		method:              http.MethodGet,
+		withRequest:         nil,
+		withResponse:        &resp,
+		acceptedStatusCodes: []int{http.StatusOK},
+		functionName:        "GetAllUpdateStatus",
+	}
+	if err := i.client.executeRequest(req); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// DefaultWaitForPendingUpdate checks each 50ms the status of a WaitForPendingUpdate.
+// This is a default implementation of WaitForPendingUpdate.
+func (i Index) DefaultWaitForPendingUpdate(updateID *AsyncUpdateID) (UpdateStatus, error) {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancelFunc()
+	return i.WaitForPendingUpdate(ctx, time.Millisecond*50, updateID)
+}
+
+// WaitForPendingUpdate waits for the end of an update.
+// The function will check by regular interval provided in parameter interval
+// the UpdateStatus. If it is not UpdateStatusEnqueued or the ctx cancelled
+// we return the UpdateStatus.
+func (i Index) WaitForPendingUpdate(
+	ctx context.Context,
+	interval time.Duration,
+	updateID *AsyncUpdateID) (UpdateStatus, error) {
+	for {
+		if err := ctx.Err(); err != nil {
+			return "", err
+		}
+		update, err := i.GetUpdateStatus(updateID.UpdateID)
+		if err != nil {
+			return UpdateStatusUnknown, nil
+		}
+		if update.Status != UpdateStatusEnqueued {
+			return update.Status, nil
+		}
+		time.Sleep(interval)
+	}
+}
+
+func (i Index) Search(request SearchRequest) (*SearchResponse, error) {
+	resp := &SearchResponse{}
+	searchPostRequestParams := map[string]interface{}{}
+	if request.Limit == 0 {
+		request.Limit = 20
+	}
+	if !request.PlaceholderSearch {
+		searchPostRequestParams["q"] = request.Query
+	}
+	if request.Filters != "" {
+		searchPostRequestParams["filters"] = request.Filters
+	}
+	if request.Offset != 0 {
+		searchPostRequestParams["offset"] = request.Offset
+	}
+	if request.Limit != 20 {
+		searchPostRequestParams["limit"] = request.Limit
+	}
+	if request.CropLength != 0 {
+		searchPostRequestParams["cropLength"] = request.CropLength
+	}
+	if len(request.AttributesToRetrieve) != 0 {
+		searchPostRequestParams["attributesToRetrieve"] = request.AttributesToRetrieve
+	}
+	if len(request.AttributesToCrop) != 0 {
+		searchPostRequestParams["attributesToCrop"] = request.AttributesToCrop
+	}
+	if len(request.AttributesToHighlight) != 0 {
+		searchPostRequestParams["attributesToHighlight"] = request.AttributesToHighlight
+	}
+	if request.Matches {
+		searchPostRequestParams["matches"] = request.Matches
+	}
+	if len(request.FacetsDistribution) != 0 {
+		searchPostRequestParams["facetsDistribution"] = request.FacetsDistribution
+	}
+	if request.FacetFilters != nil {
+		searchPostRequestParams["facetFilters"] = request.FacetFilters
+	}
```

Comment:
All of these default values were hardcoded before. I can not come up with a good solution for this, but just as a question to think about, is there maybe any other way to store these default values somewhere instead of lost in the middle of the code?

For example: Limit => 20

Reply:
@eskombro 
We could state:
``` go
const (
    DefaultAttributesToRetrieve   []string = nil
    DefaultFacetFilters.          interface{} = nil
    DefaultLimit                  int64 = 0
    // with all default values
)
```
Then use:
``` go
	if request.Limit =! DefaultLimit {
		searchPostRequestParams["limit"] = request.Limit
	}
```

And it could be more explicit in:
```
	if request.Limit == DefaultLimit {
		request.Limit = 20
	}
```
too.
I'm not convinced of this, what did you think?

Reply:
I agree, I am not a big fan, but at least is more readable for 2 reasons:

- It explains why theses values are there, by adding a name to this constants
- It regroups all of the hardcoded values at one place, probably at the beginning of the file.

This same principle could be applied by creating a hashmap, or a struct.

So even if I don't think this is an amazing solution, it still feels like an improvement :)

WDYT?

Reply:
Same! I will do the change.
Thanks again @eskombro